src/_P039_Thermosensors.ino

```diff
@@ -0,0 +1,1681 @@
+
+  # ifndef BUILD_NO_DEBUG
+
+    // FIXME TD-er: This static flag is shared among all instances of this plugin
+    static bool sensorFault = false;
+ 
+    sensorFault = (sr != 0); // Set new state
+
+    const bool faultResolved = sensorFault && (sr == 0);
+
+    if (loglevelActiveFor(LOG_LEVEL_DEBUG)) 
+    {
+      if (sensorFault || faultResolved) {
+        String log;
+        if((log.reserve(66u))) {
+          log = F("P039 : MAX31856");
+          
+          if (sensorFault == 0) {
+            log += F("Fault resolved");
+          } else {
+            log += F("Fault :");
+
+            if (sr & 0x01) {
+              log += F(" Open (no connection)");
+            }
+
+            if (sr & 0x02) {
+              log += F(" Over/Under Voltage");
+            }
+
+            if (sr & 0x04) {
+              log += F(" TC Low");
+            }
+
+            if (sr & 0x08) {
+              log += F(" TC High");
+            }
+
+            if (sr & 0x10) {
+              log += F(" CJ Low");
+            }
+
+            if (sr & 0x20) {
+              log += F(" CJ High");
+            }
+
+            if (sr & 0x40) {
+              log += F(" TC Range");
+            }
+
+            if (sr & 0x80) {
+              log += F(" CJ Range");
+            }
+          }
+          addLog(LOG_LEVEL_DEBUG, log);
+        }
+      }
+    }
+
+  # endif // ifndef BUILD_NO_DEBUG
+
+  const bool Plugin_039_SensorAttached = (sr == 0);
+
+  if (Plugin_039_SensorAttached)
+  {
+    registers[MAX31856_RAWVALUE] >>= 5; // bottom 5 bits are unused
+    // We're left with (24 - 5 =) 19 bits
+    float temperature = Plugin_039_convert_two_complement(registers[MAX31856_RAWVALUE], 19);
+
+    // Calculate Celsius
+    return temperature / 128.0f;
+  }
+  else
+  {
+    // Fault state, thus output no value.
+    return NAN;
+  }
+}
+
+int Plugin_039_convert_two_complement(uint32_t value, int nr_bits) {
+  const bool negative = (value & (1 << (nr_bits - 1))) != 0;
+  int nativeInt;
+
+  if (negative) {
+    // Add zeroes to the left to create the proper negative native-sized integer.
+    nativeInt = value | ~((1 << nr_bits) - 1);
+  } else {
+    nativeInt = value;
+  }
+  return nativeInt;
+}
+
+float readMax31865(struct EventStruct *event)
+{
+
+  uint8_t registers[MAX31865_NO_REG] = {0};
+  uint16_t rawValue = 0u;
+
+  uint8_t CS_pin_no = get_SPI_CS_Pin(event);
+
+  // clear all faults
+  clearFaults(CS_pin_no);
+
+  // set frequency filter
+  chooseFilterType(CS_pin_no, P039_RTD_FILT_TYPE);
+
+  //activate BIAS short before read, to reduce power consumption
+  handleBias(CS_pin_no, true);
+
+  // wait for external capacities to load (min. 10ms -> give 50% adder to "be sure")
+  delay(15);
+
+  // configure read access with configuration from web interface
+  setConType(CS_pin_no, P039_RTD_CON_TYPE);
+
+  //activate one shot conversion
+  startOneShotConversion(CS_pin_no);
+
+  // wait for 100ms -> conversion to be ready
+  delay(100);
```

Comment:
@chri-kai-in @TD-er consider to use TIMER API here, maybe 100ms could be used more productive

Reply:
Ah good one... Had not yet looked thoroughly through the code, but delays of longer than a few msec should be avoided.

Reply:
Valid Point - i was hesitating to go for a polling loop to identify if conversion is finished. This would in addtion create bus traffic on SPI " for nothing". I will get myself known to the API and transfer the solution.

Reply:
You can also use the ESPEasy scheduler to get back to this (100 msec is 1/10th of a second.... Hint :) ) and let the ESP address other tasks in this eternity.

Reply:
> Valid Point - i was hesitating to go for a polling loop to identify if conversion is finished. This would in addtion create bus traffic on SPI " for nothing". I will get myself known to the API and transfer the solution.

Hint for API use:
https://github.com/letscontrolit/ESPEasy/blob/52fb53ab6bcac348a62620feb883e982d5d2e156/src/_P017_PN532.ino#L205-L208
https://github.com/letscontrolit/ESPEasy/blob/52fb53ab6bcac348a62620feb883e982d5d2e156/src/_P017_PN532.ino#L120-L130

Reply:
That's a possibility, or you simply consider the sensor reading as a state machine. (see also Dallas, BME280 and lots of other plugins)
Since you apparently need to wait 100 msec, you can also try to process the states in the `PLUGIN_TEN_PER_SECOND` call (that was the hint....)

Reply:
In my opinion it could be done without state machine, just make use of scheduler timer for this one case - we pass task id.
But yes you can add FIXME or TODO comment in code

Reply:
@TD-er, @uzi18   - I will add the comment anyhow, to not forget the task. I will check this weekend if the timer option could be a quicker temporary solution than the state machine and in that case aim for an implementation. I would prefer the state machine for the long run, as it would offer more options for other devices wih deviating timing constraints - e.g. LM7x offers shutdown mechanisms which would require wakeup calls and waiting times and i presume that would not be the last time we would face such requirement. I will provide a minimum commit/push end of this weekend to close the low hanging fruits.

Reply:
I consider this solved now - that was a good step of learning for ESPEasy programming - and still so much way ahead...